util/chunk/disk.go

```diff
@@ -367,3 +372,61 @@ func (format *diskFormatRow) toMutRow(fields []*types.FieldType) MutRow {
 	}
 	return MutRow{c: chk}
 }
+
+// ReaderWithCache helps to read data that has not be flushed to underlying layer.
+// By using ReaderWithCache, user can still write data into ListInDisk even after reading.
+type ReaderWithCache struct {
+	r        io.ReaderAt
+	cacheOff int64
+	cache    []byte
+}
+
+// NewReaderWithCache returns a ReaderWithCache.
+func NewReaderWithCache(r io.ReaderAt, cache []byte, cacheOff int64) *ReaderWithCache {
+	return &ReaderWithCache{
+		r:        r,
+		cacheOff: cacheOff,
+		cache:    cache,
+	}
+}
+
+// ReadAt implements the ReadAt interface.
+func (r *ReaderWithCache) ReadAt(p []byte, off int64) (readCnt int, err error) {
+	readCnt, err = r.r.ReadAt(p, off)
+	if err != io.EOF {
+		return readCnt, err
+	}
+
+	if len(p) == readCnt {
+		return readCnt, err
+	} else if len(p) < readCnt {
+		return readCnt, errors2.Trace(errors2.Errorf("cannot read more data than user requested"+
+			"(readCnt: %v, len(p): %v", readCnt, len(p)))
+	}
+
+	// When got here, user input is not filled fully, so we need read data from cache.
+	err = nil
+	if readCnt == 0 {
+		// 'readCnt == 0' means all user requested data resides in r.cache.
+		beg := off - r.cacheOff
+		if beg < 0 {
+			return readCnt, errors2.Trace(errors2.New("off must be greater than r.cacheOff when readCnt is 0"))
+		}
+		end := int(beg) + len(p)
+		if end > len(r.cache) {
+			err = io.EOF
+			end = len(r.cache)
+		}
+		readCnt = copy(p, r.cache[beg:end])
+	} else {
+		// readCnt != 0 means only partial data of user requested resides in r.cache.
+		p = p[readCnt:]
+		end := len(p)
+		if end > len(r.cache) {
+			err = io.EOF
+			end = len(r.cache)
+		}
+		readCnt += copy(p, r.cache[:end])
+	}
+	return readCnt, err
+}
```

Comment:
why this `if` necessary? I suppose you can combine the two branch by refactoring?

Reply:
Refactored. Make the code more concise.